### PR TITLE
pybind/rados: add aio_set_xattr() and aio_rm_xattr()

### PR DESCRIPTION
**Pull request by zjcmszh**
Signed-off-by: Zhang Jiao <zhangjiao@cmss.chinamobile.com>

### CODE REVIEW

File: src/pybind/rados/rados.pyx [hunk 1]
```diff
@@ -2536,6 +2539,80 @@ cdef class Ioctx(object):
...
+
+        if ret < 0:
+            completion._cleanup()
+            raise make_ex(ret, "Failed to set xattr %r" % xattr_name)
+        return completion
+
+    @requires(('object_name', str_type), ('xattr_name', str_type), ('oncomplete', opt(Callable)))
+    def aio_rm_xattr(self, object_name, xattr_name, oncomplete = None):
+        """
+        synchronously delete an extended attribute from an object
```
Comment (ideepika) [inline src/pybind/rados/rados.pyx]: should be `s/synchronously/Asynchronously`

Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: @ideepika updated, please reveiw.

Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: updated, please review, thanks a lot.

Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: test code:
```
from rados import (Rados, WriteOpCtx, LIBRADOS_CREATE_EXCLUSIVE, NoData)
from nose.tools import eq_ as eq, ok_ as ok, assert_raises
import threading
import rados
import unittest

class TestIoctx(unittest.TestCase):
    def setUp(self):
        self.rados = Rados(conffile='./ceph.conf')
        self.rados.connect()
        self.rados.create_pool('pool002')
        assert self.rados.pool_exists('pool002')
        self.ioctx = self.rados.open_ioctx('pool002')
    def tearDown(self):
        self.ioctx.close()
        self.rados.delete_pool('pool002')
        self.rados.shutdown()

    def test_aio_set_xattr(self):
        lock = threading.Condition()
        count = [0]
        def cb(blah):
            with lock:
                count[0] += 1
                lock.notify()
            return 0
        comp = self.ioctx.aio_set_xattr("obj", "key", "value", cb)
        comp.wait_for_complete()
        with lock:
            while count[0] < 1:
                lock.wait()
        eq(comp.get_return_value(), 0)
        eq(self.ioctx.get_xattr('obj', 'key'), 'value')

    def test_aio_rm_xattr(self):
        lock = threading.Condition()
        count = [0]
        def cb(blah):
            with lock:
                count[0] += 1
                lock.notify()
            return 0
        self.ioctx.set_xattr('xyz', 'key', 'value')
        eq(self.ioctx.get_xattr('xyz', 'key'), 'value')
        comp = self.ioctx.aio_rm_xattr('xyz', "key", cb)
        comp.wait_for_complete()
        with lock:
            while count[0] < 1:
                lock.wait()
        eq(comp.get_return_value(), 0)
        with assert_raises(NoData):
            self.ioctx.get_xattr('xyz', 'key')

if __name__ == '__main__':
    unittest.main()
```
test results:
```
[root@jiaowa build]# python test_aio_xttrs.py
..
----------------------------------------------------------------------
Ran 2 tests in 5.792s

OK
```
please check.

File: src/pybind/rados/rados.pyx [hunk 2]
```diff
@@ -2536,6 +2539,80 @@ cdef class Ioctx(object):
             raise make_ex(ret, "error reading %s" % object_name)
         return completion
 
+    @requires(('object_name', str_type), ('xattr_name', str_type),
+              ('xattr_value', bytes), ('oncomplete', opt(Callable)))
+    def aio_set_xattr(self, object_name, xattr_name, xattr_value, oncomplete = None):
```
Comment (ideepika) [inline src/pybind/rados/rados.pyx]: please consider changing `aio_set_xattr` to `aio_setxattr` looking into librados code, that seems to be more relevant, I'd suggest following the same convention in all your work. Thanks.

File: src/test/pybind/test_rados.py [hunk 3]
```diff
@@ -739,6 +739,41 @@ def cb(blah):
...
 
+    def test_aio_set_xattr(self):
+        lock = threading.Condition()
+        count = [0]
+        def cb(blah):
+            with lock:
+                count[0] += 1
+                lock.notify()
+            return 0
+        comp = self.ioctx.aio_set_xattr("obj", "key", "value", cb)
```
Comment (ideepika) [inline src/test/pybind/test_rados.py]: think we should use a byte literal for `value`.

File: src/pybind/rados/rados.pyx (hunk 2 above)
Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]:
```
    @requires(('xattr_name', str_type), ('xattr_value', bytes))
    def set_xattr(self, xattr_name, xattr_value):
        """
        Set an xattr
        :param xattr_name : name of the xattr
        :type xattr_name : strgg
        :param xattr_value: buffer to set xattr to
        :type xattr_value: bytes
        """
        xattr_name = cstr(xattr_name, 'xattr_name')
        cdef:
            char *_xattr_name = xattr_name
            char *_xattr_value = xattr_value
            size_t _xattr_value_len = len(xattr_value)
        with nogil:
            rados_write_op_setxattr(self.write_op, _xattr_name, _xattr_value, _xattr_value_len)
```
this is my previous commit,you recommend me to chang setxattr() to set_xattr().
 I'm so confused today.

Comment (ideepika) [reply src/pybind/rados/rados.pyx]: apologies for the same, but taking a closer look into source it'd be more consistent to have the same name of function in python and C.

File: src/test/pybind/test_rados.py (hunk 3 above)
Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: @ideepika updated please check.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2019-12-09_12:39:45-rados-wip-kefu-testing-2019-12-09-1434-distro-basic-smithi/

failure tracked / fixed by
- https://tracker.ceph.com/issues/43192
- https://tracker.ceph.com/issues/42764
- https://tracker.ceph.com/issues/43070
- #32143

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by zjcmszh** [comment]
If there is no problem, can you merge it as soon as possible ? thank you very very very  very very much

**Comment by zjcmszh** [comment]
@tchaikov  If there is no problem, can you merge it as soon as possible ? thank you very very very very very much

**Pull request closed without merging**